### PR TITLE
gh-95273: Reorganize sqlite3 doc module level funcs and vars

### PR DESCRIPTION
**Pull request by erlend-aasland**
Put module level functions before module level attributes, and (mostly)
sort them alphabetically.  This makes sqlite3.connect() the first
encounter in the sqlite3 reference.

* Issue: gh-95273

### CODE REVIEW

**Comment by erlend-aasland** [comment]
Other things to note, aside of the nice `sqlite3.connect` promotion:

- `sqlite3.version*` is now at the end; that's ok, since they're confusing to people, and they are deprecated :)
- it's _almost_ alphabetically sorted, but we cannot start with the obscure `sqlite3.complete_statement`[^1]

[^1]: ... which has a far too large example for being such an obscure function; IMO, we can remove that example (or at least greatly simplify it), or maybe just point to the source code of the newly added sqlite3 CLI.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
.. function:: connect(database, timeout=5.0, detect_types=0, \
                      isolation_level="DEFERRED", check_same_thread=True, \
                      factory=sqlite3.Connection, cached_statements=128, uri=False)
```

You need to escape the line break or it breaks the syntax and neither resolves the `:func:`s nor renders the function correctly; see [the Sphinx docs](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#basic-markup) for more details.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]: Maybe worth splitting this into two sections, one for functions and one for module-level constants, for ease of identification and navigation, especially since `UPPER_SNAKE_CASE` isn't used consistently to identify constants, and it makes the alphabetical ordering more obvious?

File: Doc/library/sqlite3.rst
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst:142]: Do you think it would be useful to do the following?
1. add sub-sections for functions and constants
2. add a sentence like ``This section lists :mod:`!sqlite3`'s :ref:`module-functions` and :ref:`module-constants`.``

**Review by CAM-Gerlach — CHANGES_REQUESTED** [review]
Great idea overall, this is certainly a significant improvement. There is one serious syntax issue that needs to be fixed, and I wonder whether it might be a better idea to split the module-level constants and functions into two separate sections for clarity.

**Review by ezio-melotti — COMMENTED** [review]
The changes LGTM, assuming you just moved things around and wrapped a few long lines.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst:142]: > 1\. add sub-sections for functions and constants

:+1: , I also suggested the same thing above, heh.

> 2\. add a sentence like `` This section lists :mod:`!sqlite3`'s :ref:`module-functions` and :ref:`module-constants`. ``

Especially if we do the previous, I'm not sure we really need this—it just restates what is already in the section titles and in the ToC in the sidebar. Did you mean it as an alternative?

Comment (ezio-melotti) [reply Doc/library/sqlite3.rst:142]: > +1 , I also suggested the same thing above, heh.

Great minds think alike ;)

> Especially if we do the previous, I'm not sure we really need this—it just restates what is already in the section titles and in the ToC in the sidebar. Did you mean it as an alternative?

The idea was both to set expectations on the content of the section (even though the title already does that), and also to provide a link to skip the list of function and jump to the constants without having to manually scroll through.

I was also thinking if this is something we should start doing/recommending in the style guide too :thinking: The sidebar already provides links, but after reaching  the "Module functions and constants" section, it kind of breaks the flow to switch over to the sidebar to find out which section we are in and see/click on the desired sub-section (some themes might make this easier though).  Normally I think I would just scroll through, but if there was a handy inline link I would probably click it.

Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst:142]: > Great minds think alike ;)

Funny enough, I actually typed out a reference to the same phrase out above, but decided against it before posting (since my mind isn't anywhere near as great as yours) :laughing: 

> The idea was both to set expectations on the content of the section (even though the title already does that), and also to provide a link to skip the list of function and jump to the constants without having to manually scroll through.

Hmm...this is something I do in the [spec section of PEP 639](https://peps.python.org/pep-0639/#specification) (perhaps with excessive detail), but I'm not sure about in the API reference. Unlike a draft PEP, people aren't generally scrolling through it, they rather search or heading jump to specifically what they are looking for. And while "flow" is important in a tutorial, it is not so much in a reference. Per [Diataxis](https://diataxis.fr/reference/#reference-as-description),

> Reference material should be austere and to the point. One hardly *reads* reference material; one *consults* it.

Still, if there was a nested hierarchy of sections, this could still certainly be valuable by providing a concise and cross-referenced summary at the top level, but given all of the sections are standalone underneath the reference, I'm not sure this one-off case necessarily is worth it—particularly under my suggestion which just splits them into two separate sections, since there really isn't much shared commonality between them.

Like with other organizational things this could be something to think about for the future, but if I'd say we discuss and try it out first to see how it works before formally codifying it, if we do.

Comment (ezio-melotti) [reply Doc/library/sqlite3.rst:142]: Fair enough :+1: 

> Unlike a draft PEP, people aren't generally scrolling through it, they rather search or heading jump to specifically what they are looking for.

To clarify, my thinking is that people will either ctrl+f or jump to the first header that sounds close to what they are looking for (e.g. "Module functions and constants", rather than the specific sub-section) and once they got close enough they will (have to) scroll through.  I don't expect them to scroll through the whole page.

> I'm not sure this one-off case necessarily is worth it—particularly under my suggestion which just splits them into two separate sections

Agreed.  If this is split into two separate top-level section there is no need (nor place) for an additional sentence that links to functions and constants.  In addition, having these at the top level will make them easier to find/notice.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:142]: FTR, _"Module functions and constants"_ is a **subsubsection**[^1], not a **section** or a **subsection**. I'm a little confused if you imply we should create two _subsubsections_ or keep one subsubsection with two _paragraphs_[^2].

I _assume_ you mean the latter :)

[^1]: Subsubsections use `^^^^^`
[^2]: Paragraphs use `"""""`

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:142]: Also, I think _Module functions and attributes_ is better than _Module functions and constants_.

Comment (ezio-melotti) [reply Doc/library/sqlite3.rst:142]: The two options are:
1. Split "Module functions and constants" into "Module functions" and "Module constants", both at the same level where "Module functions and constants" is now (so they should be subsubsections with `^^^^^`)
2. Keep "Module functions and constants", and add  "Module functions" and "Module constants" nested inside (so as paragraphs with `"""""`)

Option number 1 is probably better (flat is better than nested).

> Also, I think Module functions and attributes is better than Module functions and constants.

I generally prefer to use "attributes" for classes/instances, and I would generally refer to this as "module-level variables" or "global variables" (even though "global" is somewhat ambiguous).  Constants it technically inaccurate, since they are only constants in spirit, but they are actually variables.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:142]: > Option number 1 is probably better (flat is better than nested).

+1

> I generally prefer to use "attributes" for classes/instances, and I would generally refer to this as "module-level variables" or "global variables" (even though "global" is somewhat ambiguous). Constants it technically inaccurate, since they are only constants in spirit, but they are actually variables.

I don't like "constants" either. ~Personally, I'd lean towards "attributes" instead of "variables", but~ I have no strong opinion here; ~either~ "variables" is fine by me.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:142]: PTAL :)

File: Doc/library/sqlite3.rst
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst]:
```suggestion
   :ref:`type that SQLite natively understands <sqlite3-types>`.
```

**Review by ezio-melotti — APPROVED** [review]
A small inconsequential nit, otherwise LGTM.
I've seen refs without spaces elsewhere, so you might want to double-check the whole file, possibly in a separate PR (perhaps together with adding `!` to `` `:mod:`sqlite3` ``).

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst:142]: For what its worth, I prefer "module-level constants" since that describes their intended semantics, even if it is of course technically possible modify them, just like we call "private" attributes as such, and the like. [PEP 8 calls them such as well](https://peps.python.org/pep-0008/#constants). But I'm not a Python expert or anything like you two so take that for what you will :laughing:

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
   For example:

   .. code-block:: sql
```

Nit: Extra line break here

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
   case-insensitively.


.. _sqlite3-module-variables:
```

Two spaces after headings, no?

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst:142]: And I'm not sure if other modules use this terminology? Or maybe they do...

**Review by CAM-Gerlach — APPROVED** [review]
LGTM, just a couple tiny nits.

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:142]: > And I'm not sure if other modules use this terminology? Or maybe they do...

I'll check current practice before proceeding with landing this.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:142]: Current practices are:

- Only `logging` has a heading that start with _Module level ..._; everyone else use _Module ..._
- "Constants" is used for upper-case module level attributes (most often used)
- "Attributes" is used for lower-case module level attributes (seldom used)
- "Variables" is not used for module level attributes

Also, note that a lot of modules describe exceptions in their _Module ..._ subsections.

Since the sqlite3 module has a mix of upper-case and lower-case module level attributes, we could either split them up in _Module attributes_ and _Module constants_, or just use one of them (perhaps _Module constants_).

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:142]: > Since the sqlite3 module has a mix of upper-case and lower-case module level attributes, we could either split them up in _Module attributes_ and _Module constants_, or just use one of them (perhaps _Module constants_).

I used _Module constants_. I'm marking this as resolved.

File: Doc/library/sqlite3.rst
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst:287]: If you want, you could add a sentence that introduces the constants defined in the module, possibly mentioning the fact that their values shouldn't be changed even though it's technically possible to do it.

**Review by ezio-melotti — APPROVED** [review]
One last suggestion, but otherwise LGTM.

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:287]: > If you want, you could add a sentence that introduces the constants defined in the module [...]

My initial reaction is -0. I'll see if I can phrase something meaningful, though; if I can't, I won't add anything :)

> [...] possibly mentioning the fact that their values shouldn't be changed even though it's technically possible to do it.

I'm -1
No other module takes this approach, and IMO, adding such a sentence is kinda like pointing at the red button labelled _DO NOT PRESS_ 🔴 😄

Comment (ezio-melotti) [reply Doc/library/sqlite3.rst:287]: True, perhaps the sentence is unnecessary.  The reason why I suggested it is that the title is potentially misleading since it seems to imply that these are actually constants -- the sentence would help clarify that they aren't.  This is just nitpicking though, feel free to merge without extra sentences :)

**Pull request merged**
